sips/sip-015/sip-015-network-upgrade.md

```diff
@@ -0,0 +1,1644 @@
+services, and to implement "transaction-less transactions.""  The ability to
+both verify the signature on a buffer and decode the buffer into a Clarity
+value is a basic building block for bridges and oracles.  In addition, it is a
+basic building block for smart contracts where users are not expected to send
+transactions (e.g. because they lack a wallet, or because the transaction volume
+would be too high).  Instead, they can send signed Clarity values to someone willing to
+package them into transactions at a later time (possibly as a batch), and
+send them to the smart contract on their behalf.
+
+**Examples:**
+
+```clarity
+(from-consensus-buff int 0x0000000000000000000000000000000001) ;; Returns (some 1)
+(from-consensus-buff uint 0x0000000000000000000000000000000001) ;; Returns none
+(from-consensus-buff uint 0x0100000000000000000000000000000001) ;; Returns (some u1)
+(from-consensus-buff bool 0x0000000000000000000000000000000001) ;; Returns none
+(from-consensus-buff bool 0x03) ;; Returns (some true)
+(from-consensus-buff bool 0x04) ;; Returns (some false)
+(from-consensus-buff principal 0x051fa46ff88886c2ef9762d970b4d2c63678835bd39d) ;; Returns (some SZ2J6ZY48GV1EZ5V2V5RB9MP66SW86PYKKQ9H6DPR)
+(from-consensus-buff { abc: int, def: int } 0x0c00000002036162630000000000000000000000000000000003036465660000000000000000000000000000000004) ;; Returns (some (tuple (abc 3) (def 4)))
+```
+
+### New method: `replace-at`
+
+* **Input Signature:** `(replace-at (x (sequence Y)) (i uint) (value Y))`
+* **Output Signature:** `(optional (sequence Y))`
+
+The `replace-at` function takes in a sequence, an index, and an element, 
+and returns a new sequence with the data at the index position replaced with the given element. 
+The given element's type must match the type of the sequence, and must correspond to a single 
+index of the input sequence. The return type on success is the same type as the input sequence. 
+The sequence can be a list, buffer, string-ascii, or string-utf8.
+
+If the provided index is out of bounds, this functions returns `none`.
+
+Note that sequences are zero-indexed.  The first item in `x` is at index 0.
+
+**Rationale:** This method makes it possible to set values of a sequence type
+without incurring the costs of splitting the sequence into subsequences,
+and concatenating them back together with the new value.  This operation is also
+a common facility in many other languages, so supporting it in Clarity would
+make Clarity more approachable to new developers.
+
+**Examples:**
+
+```clarity
+(replace-at u"ab" u1 u"c") ;; Returns (some u"ac")
+(replace-at 0x00112233 u2 0x44) ;; Returns (some 0x00114433)
+(replace-at "abcd" u3 "e") ;; Returns (some "abce")
+(replace-at (list 1) u0 10) ;; Returns (some (10))
+(replace-at (list (list 1) (list 2)) u0 (list 33)) ;; Returns (some ((33) (2)))
+(replace-at (list 1 2) u3 4) ;; Returns none
+```
+
+### New global: `tx-sponsor?`
```

Comment:
Drop the question mark.
All other keywords have plain kebab-case names, and this one should follow the same standard.

Clarity already have a mix of functions that use `!` and `?` in a inconsistent way. Do not apply same to keywords. Keep it simple.

Reply:
As @LNow on mentioned on discord, the following functions should have a `?`:


principal-destruct, principal-construct - both return (response X Y), and both can return success, semi-error and error
slice - return optional
string-to-int - optional
string-to-uint - optional
to-consensus-buff - optional
from-consensus-buff - optional
replace-at - optional

Furthermore, `element-at?` and `index-of?` should be added as new functions with the same behaviour of the functions without the question mark.

Reply:
I'll open a PR for this implementation. Any problems with this?

Reply:
See https://github.com/stacks-network/stacks-blockchain/pull/3374